File: tooling/node/support.mjs
// Synchronous capstone doubles; these import no challenge implementations.
export class FakeStore {
  entries = new Map();
  constructor(clock) { this.clock = clock; }
  get(key) {
    const entry = this.entries.get(key);
    return entry !== undefined && this.clock() < entry.deadline ? entry.value : null;
  }
  put(key, value, ttlMs) {
    this.entries.set(key, { value, deadline: this.clock() + ttlMs });
  }
}

export class FakeQueue {
  ready = [];
  active = new Map();
  dead = [];
  nextJob = 1;
  nextLease = 1;
  constructor(maxAttempts) { this.maxAttempts = maxAttempts; }
  enqueue(payload) {
    const id = "job-" + this.nextJob++;
    this.ready.push({ job_id: id, payload, attempts: 0 });
    return id;
  }
  reserve() {
    const job = this.ready.shift();
    if (job === undefined) return null;
    job.attempts++;
    const receipt = "lease-" + this.nextLease++;
    this.active.set(receipt, job);
    return { job_id: job.job_id, payload: job.payload, attempt: job.attempts, receipt };
  }
  ack(receipt) { return this.active.delete(receipt); }
  nack(receipt) {
    const job = this.active.get(receipt);
    if (job === undefined) return false;
    this.active.delete(receipt);
    if (job.attempts >= this.maxAttempts) this.dead.push({ ...job });
    else this.ready.push(job);
    return true;
  }
}

export class FakeSender {
  attempts = [];
  sent = [];
  constructor(outcomes) { this.outcomes = [...outcomes]; }
  send(message) {
    this.attempts.push(message);
    const outcome = this.outcomes.shift() ?? "ok";
    if (outcome === "raise") throw new Error("Scripted sender failure");
    if (outcome === "fail") return false;
    this.sent.push(message);
    return true;
  }
}

File: tooling/python/support.py
"""Small, synchronous capstone doubles. No earlier exercise is imported."""
from collections import deque


class FakeStore:
    def __init__(self, clock):
        self.clock = clock
        self.entries = {}

    def get(self, key):
        entry = self.entries.get(key)
        return entry[0] if entry is not None and self.clock() < entry[1] else None

    def put(self, key, value, ttl_ms):
        self.entries[key] = (value, self.clock() + ttl_ms)


class FakeQueue:
    """Only synchronous reserve/ack/nack; deliberately no visibility scheduler."""
    def __init__(self, max_attempts):
        self.max_attempts = max_attempts
        self.ready = deque()
        self.active = {}
        self.dead = []
        self.next_job = 1
        self.next_lease = 1

    def enqueue(self, payload):
        job_id = f"job-{self.next_job}"
        self.next_job += 1
        self.ready.append({"job_id": job_id, "payload": payload, "attempts": 0})
        return job_id

    def reserve(self):
        if not self.ready:
            return None
        job = self.ready.popleft()
        job["attempts"] += 1
        receipt = f"lease-{self.next_lease}"
        self.next_lease += 1
        self.active[receipt] = job
        return {"job_id": job["job_id"], "payload": job["payload"],
                "attempt": job["attempts"], "receipt": receipt}

    def ack(self, receipt):
        return self.active.pop(receipt, None) is not None

    def nack(self, receipt):
        job = self.active.pop(receipt, None)
        if job is None:
            return False
        if job["attempts"] >= self.max_attempts:
            self.dead.append(job)
        else:
            self.ready.append(job)
        return True


class FakeSender:
    def __init__(self, outcomes):
        self.outcomes = deque(outcomes)
        self.attempts = []
        self.sent = []

    def __call__(self, message):
        self.attempts.append(message)
        outcome = self.outcomes.popleft() if self.outcomes else "ok"
        if outcome == "raise":
            raise RuntimeError("Scripted sender failure")
        if outcome == "fail":
            return False
        self.sent.append(message)
        return True
